flink-ml-lib/src/main/java/org/apache/flink/ml/feature/lsh/LSHModel.java

```diff
@@ -373,34 +395,45 @@ public TopKFunction(String distCol, int numNearestNeighbors) {
         }
 
         @Override
-        public PriorityQueue<Row> createAccumulator() {
-            return new PriorityQueue<>(numNearestNeighbors, new DistColComparator(distCol));
+        public List<Row> createAccumulator() {
+            return new ArrayList<>(numNearestNeighbors);
         }
 
         @Override
-        public PriorityQueue<Row> add(Row value, PriorityQueue<Row> accumulator) {
-            if (accumulator.size() == numNearestNeighbors) {
-                Row peek = accumulator.peek();
-                if (accumulator.comparator().compare(value, peek) < 0) {
-                    accumulator.poll();
-                }
+        public List<Row> add(Row value, List<Row> accumulator) {
+            if (topKRows == null) {
+                topKRows = new PriorityQueue<>(numNearestNeighbors, new DistColComparator(distCol));
+                topKRows.addAll(accumulator);
             }
-            accumulator.add(value);
-            return accumulator;
+            insert(value, topKRows);
+            return new ArrayList<>(topKRows);
```

Comment:
It would be pretty inefficient to copy the entire list every time we add a value.

How about we use `PriorityQueue<Row>` as the accumulator and let `getResult(List<Row> accumulator)` return `List<ROW>`? I think we will only need to serialize the output of `getResult` but not the accumulator itself.

Reply:
Note that if the aggregator needs to be serializable by either `kryo` or our custom serializer, we can define a specific container class that has the `PriorityQueue<ROW>` as member variable and define a custom serializer for it.